### PR TITLE
Support `STATE_CHANGES` sidecar

### LINKED ISSUE #7358: Restore `STATE_CHANGES` sidecars

### PR DESCRIPTION
**Pull request by tinker-michaelj**
**Description**:
 - Closes #7358 
 - Adds a [`StorageAccessTracker`](https://github.com/hashgraph/hedera-services/blob/07358-support-state-changes/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/infra/StorageAccessTracker.java#L31) to encapsulate logic for tracking and summarizing storage access, both reads and writes.
    * Note that _during_ the EVM transaction, we only track reads (`SLOAD`s); and then when committing a successful transaction, merge in the updates from any `SSTORE`s that were done.
 - Implements the [`CustomSLoadOperation` ](https://github.com/hashgraph/hedera-services/blob/07358-support-state-changes/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/operations/CustomSLoadOperation.java#L38)to wrap the Besu `SLoadOperation` and track `SLOAD`'s via the `StorageAccessTracker` in the frame's context (if available).
 - When preparing the success result for an initial frame with a `StorageAccessTracker`, merges the [pending updates with the tracked reads](https://github.com/hashgraph/hedera-services/blob/35f9e980fbb9f6e1679f7872a6edef6aaad95a5c/hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransactionResult.java#L150) and includes the net accesses as a PBJ `ContractStateChanges` in the result.

### CODE REVIEW

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/utils/FrameRunner.java [hunk 1]
```diff
@@ -86,21 +88,16 @@ public HederaEvmTransactionResult runToCompletion(
...
-    private void requireAllNonNull(
-            @NonNull final MessageFrame frame,
-            @NonNull final HederaTracer tracer,
-            @NonNull final CustomMessageCallProcessor messageCall,
-            @NonNull final ContractCreationProcessor contractCreation) {
-        requireNonNull(frame);
-        requireNonNull(tracer);
-        requireNonNull(messageCall);
-        requireNonNull(contractCreation);
+    private ContractID resolvedHederaId(@NonNull final MessageFrame frame, @NonNull final Address address) {
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/utils/FrameRunner.java:97]: TY!

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransactionResult.java [hunk 2]
```diff
@@ -16,13 +16,17 @@
 
 package com.hedera.node.app.service.contract.impl.hevm;
 
-import static com.hedera.node.app.service.contract.impl.utils.ConversionUtils.pbjLogsFrom;
-import static com.hedera.node.app.service.contract.impl.utils.ConversionUtils.tuweniToPbjBytes;
+import static com.hedera.node.app.service.contract.impl.exec.utils.FrameUtils.accessTrackerFor;
+import static com.hedera.node.app.service.contract.impl.exec.utils.FrameUtils.proxyUpdaterFor;
+import static com.hedera.node.app.service.contract.impl.utils.ConversionUtils.*;
+import static com.hedera.node.app.service.contract.impl.utils.ConversionUtils.asPbjStateChanges;
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransactionResult.java:22]: Needed with `*` import above?  (Do we even use `*` imports?)

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransactionResult.java [hunk 3]
```diff
@@ -127,10 +141,37 @@ public static HederaEvmTransactionResult resourceExhaustionFrom(
...
+        final var accessTracker = accessTrackerFor(frame);
+        if (accessTracker == null) {
+            return null;
+        } else {
+            final List<StorageAccesses> accesses;
+            if (includeWrites) {
+                final var worldUpdater = proxyUpdaterFor(frame);
+                accesses = accessTracker.getReadsMergedWith(worldUpdater.pendingStorageUpdates());
+            } else {
+                accesses = accessTracker.getJustReads();
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransactionResult.java:172]: Are these few lines worth moving to `StorageAccessTracker`?

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/infra/StorageAccessTracker.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+     * includes both the given list of writes, and all tracked first reads.
+     *
+     * <p>If a key was both read and overwritten, the read-only access will not appear,
+     * since a write {@link com.hedera.node.app.service.contract.impl.state.StorageAccess}
+     * already includes the overwritten value.
+     *
+     * @param writes all the storage
+     * @return the merged list of all storage accesses
+     */
+    public List<StorageAccesses> getReadsMergedWith(@NonNull final List<StorageAccesses> writes) {
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/infra/StorageAccessTracker.java:72]: Nicely compact/terse.  Needs a bit of concentration to read it but it all becomes clear ...

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/state/ProxyWorldUpdater.java [hunk 5]
```diff
@@ -300,7 +305,7 @@ public void commit() {
      */
     @Override
     public @NonNull WorldUpdater updater() {
-        return new ProxyWorldUpdater(scope, evmFrameStateFactory, this);
+        return new ProxyWorldUpdater(scope.begin(), evmFrameStateFactory, this);
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/state/ProxyWorldUpdater.java:308]: Easily omitted ...

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/state/StorageAccess.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    public boolean isInsertion() {
+        return writtenValue != null && !writtenValue.isZero() && value.isZero();
+    }
+
+    public boolean isReadOnly() {
+        return writtenValue == null;
+    }
+}
```
Comment (david-bakin-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/state/StorageAccess.java:60]: I guess you don't anticipate a need for `isUpdate()`?

**Review by david-bakin-sl — APPROVED** [review]
LGTM!  Almost 100% code coverage via unit tests too, much admired!  (Only things missing unit test coverage are unimplemented methods which throw exceptions and a few edge cases in `ConversionUtils.java` - which you could address at your leisure ...)

**Comment by david-bakin-sl** [comment]
A lot of plumbing.  Can't wait for action sidecars ...

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransactionResult.java (hunk 2 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransactionResult.java:22]: Fixed.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransactionResult.java (hunk 3 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransactionResult.java:172]: What method signature did you have in mind? I thought a bit and my ideas seemed awkward.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/state/StorageAccess.java (hunk 6 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/state/StorageAccess.java:99]: Added for completeness and sake of avoiding double-negatives.

**Comment by tinker-michaelj** [comment]
> A lot of plumbing. Can't wait for action sidecars ...

Indeed...starting on that tomorrow. 😋

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/state/ProxyWorldUpdater.java (hunk 5 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/state/ProxyWorldUpdater.java:308]: I mean the `begin()` you added.  Type-checks without it.  Easily missed.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransactionResult.java (hunk 3 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/hevm/HederaEvmTransactionResult.java:173]: Dunno.  Maybe something like
```
enum Updates { READS_ONLY, READS_AND_WRITES };
getStateChanges(Updates,WorldUpdater)
```

????
